File: tests/test_command_session_switch.py
# pyright: standard

import json
from pathlib import Path

from typer.testing import CliRunner

from aico.main import app

runner = CliRunner()


def test_session_switch_switches_active_pointer(tmp_path: Path) -> None:
    with runner.isolated_filesystem(temp_dir=tmp_path) as td:
        # GIVEN a shared-history project with a second session
        result_init = runner.invoke(app, ["init"])
        assert result_init.exit_code == 0

        result_new = runner.invoke(app, ["session-new", "feature"])
        assert result_new.exit_code == 0

        pointer_file = Path(td) / ".ai_session.json"
        pointer_data = json.loads(pointer_file.read_text())
        assert pointer_data["path"] == ".aico/sessions/feature.json"

        # WHEN I switch back to main
        result_switch_main = runner.invoke(app, ["session-switch", "main"])
        # THEN the pointer updates and a success message is printed
        assert result_switch_main.exit_code == 0
        assert "Switched active session to: main" in result_switch_main.stdout
        pointer_data = json.loads(pointer_file.read_text())
        assert pointer_data["path"] == ".aico/sessions/main.json"

        # WHEN I switch to feature again
        result_switch_feature = runner.invoke(app, ["session-switch", "feature"])
        # THEN the pointer updates and a success message is printed
        assert result_switch_feature.exit_code == 0
        assert "Switched active session to: feature" in result_switch_feature.stdout
        pointer_data = json.loads(pointer_file.read_text())
        assert pointer_data["path"] == ".aico/sessions/feature.json"


def test_session_switch_fails_for_missing_view(tmp_path: Path) -> None:
    with runner.isolated_filesystem(temp_dir=tmp_path):
        # GIVEN a project with only the default 'main' view
        result_init = runner.invoke(app, ["init"])
        assert result_init.exit_code == 0

        # WHEN I try to switch to a non-existent view
        result = runner.invoke(app, ["session-switch", "nope"])

        # THEN the command fails with a clear error
        assert result.exit_code == 1
        assert "Error: Session view 'nope' not found at" in result.stderr

File: tests/test_command_session_switch.rs
use assert_cmd::cargo_bin_cmd;
use std::fs;
use tempfile::tempdir;

#[test]
fn test_session_switch_switches_active_pointer() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // GIVEN a shared-history project with a second session
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["session-new", "feature"])
        .assert()
        .success();

    // WHEN I switch back to main
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["session-switch", "main"])
        .assert()
        .success();

    // THEN the pointer updates
    let pointer_path = root.join(".ai_session.json");
    let pointer_content = fs::read_to_string(pointer_path).unwrap();
    assert!(pointer_content.contains(".aico/sessions/main.json"));

    // WHEN I switch to feature again
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["session-switch", "feature"])
        .assert()
        .success();
    let pointer_content_after = fs::read_to_string(root.join(".ai_session.json")).unwrap();
    assert!(pointer_content_after.contains(".aico/sessions/feature.json"));
}

#[test]
fn test_session_switch_fails_for_missing_view() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // GIVEN a project with only default view
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("init")
        .assert()
        .success();

    // WHEN I try to switch to a non-existent view
    let cmd = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["session-switch", "nope"])
        .assert()
        .failure();

    // THEN the command fails with a clear error
    let stderr = String::from_utf8(cmd.get_output().stderr.clone()).unwrap();
    assert!(stderr.contains("Session view 'nope' not found"));
}
